spec/client/ics-006-solo-machine-client/README.md

```diff
@@ -127,205 +147,178 @@ function latestClientHeight(clientState: ClientState): uint64 {
 
 ```typescript
-function checkValidityAndUpdateState(
-  clientState: ClientState,
-  header: Header) {
-  assert(header.sequence === clientState.consensusState.sequence)
-  assert(header.timestamp >= clientstate.consensusState.timestamp)
-  assert(checkSignature(header.newPublicKey, header.sequence, header.diversifier, header.signature))
-  clientState.consensusState.publicKey = header.newPublicKey
-  clientState.consensusState.diversifier = header.newDiversifier
-  clientState.consensusState.timestamp = header.timestamp
-  clientState.consensusState.sequence++
+function  verifyClientMessage(clientMsg: ClientMessage) {
+  switch typeof(ClientMessage) {
+    case Header:
+      verifyHeader(clientMessage)
+    // misbehaviour only suppported for current public key and diversifier on solomachine
+    case Misbehaviour:
+      verifyMisbehaviour(clientMessage)
+  }
 }
-```
-
-### Misbehaviour predicate
 
-Any duplicate signature on different messages by the current public key freezes a solo machine client.
+function verifyHeader(header: header) {
+    assert(header.timestamp >= clientstate.consensusState.timestamp)
+    headerData = {
+      NewPublicKey: header.newPublicKey,
+      NewDiversifier: header.newDiversifier,
+    }
+    sigBytes = SignBytes(
+      Sequence: clientState.consensusState.sequence,
+      Timestamp: header.timestamp,
+      Diversifier: clientState.consensusState.diversifier,
+      Path: []byte{"solomachine:header"}
+      Value: marshal(headerData)
+    )
+    assert(checkSignature(cs.consensusState.publicKey, sigBytes, header.signature))
+}
 
-```typescript
-function checkMisbehaviourAndUpdateState(
-  clientState: ClientState,
-  misbehaviour: Misbehaviour) {
-    h1 = misbehaviour.h1
-    h2 = misbehaviour.h2
+function verifyMisbehaviour(misbehaviour: Misbehaviour) {
+    s1 = misbehaviour.signatureOne
+    s2 = misbehaviour.signatureTwo
     pubkey = clientState.consensusState.publicKey
     diversifier = clientState.consensusState.diversifier
     timestamp = clientState.consensusState.timestamp
     // assert that timestamp could have fooled the light client
-    assert(misbehaviour.h1.signature.timestamp >= timestamp)
-    assert(misbehaviour.h2.signature.timestamp >= timestamp)
-    // assert that signature data is different
-    assert(misbehaviour.h1.signature.data !== misbehaviour.h2.signature.data)
-    // assert that the signatures validate
-    assert(checkSignature(pubkey, misbehaviour.sequence, diversifier, misbehaviour.h1.signature.data))
-    assert(checkSignature(pubkey, misbehaviour.sequence, diversifier, misbehaviour.h2.signature.data))
-    // freeze the client
-    clientState.frozen = true
+    assert(misbehaviour.s1.timestamp >= timestamp)
+    assert(misbehaviour.s2.timestamp >= timestamp)
```

Comment:
I'm a bit confused by this assertion. The comment says:
```
// assert that timestamp could have fooled the light client
```

It's possible the misbehaviour sequence has already been processed and thus the misbehaviour timestamp could be less than the current timestamp. In this situation, either misbehaviour would have fooled the client but this check would fail

Reply:
Yea, i think this is wrong. will remove

Reply:
It would be a good check if we were checking against the consensus state of the previous sequence. But that's not possible since solomachine only stores latest consensus state